File: apps/notesprout_android/app/src/main/kotlin/com/notesprout/android/recognition/HandwritingRecognizer.kt
package com.notesprout.android.recognition

import android.graphics.RectF
import com.notesprout.android.data.LiveStroke

/**
 * General-purpose handwriting recognition interface.
 * Operates on raw stroke data with no knowledge of the calling context
 * (headings, text boxes, search indexing, etc.).
 *
 * Implementations:
 *   - MlKitHandwritingRecognizer (Google ML Kit, all devices)
 *   - Future: OnyxHwrHandwritingRecognizer (BOOX firmware AIDL bridge)
 */
interface HandwritingRecognizer : AutoCloseable {

    /**
     * Returns true if this recognizer is ready to process strokes.
     * May return false if the ML model has not been downloaded yet.
     */
    fun isReady(): Boolean

    /**
     * Recognize handwritten text from a list of strokes within a bounding box.
     *
     * @param strokes  The strokes to recognize. Must not be empty.
     * @param bounds   The writing area bounding box (used to improve accuracy).
     * @param onResult Called on the main thread with the recognized string,
     *                 or with FALLBACK_TEXT if recognition fails or the
     *                 recognizer is not ready.
     */
    fun recognize(
        strokes: List<LiveStroke>,
        bounds: RectF,
        onResult: (String) -> Unit
    )

    /**
     * Recognize a single text line, feeding [preContext] (the previously recognized line)
     * into the model for per-line context chaining — the single biggest free accuracy win
     * for page-scale recognition. Suspends until the model returns.
     *
     * Returns the recognized string, or [FALLBACK_TEXT] if recognition fails or the
     * recognizer is not ready. Safe to call off the main thread.
     *
     * Used by the page/notebook text pipeline (StrokeSegmenter → PageTextRecognizer).
     */
    suspend fun recognizeSegment(
        strokes: List<LiveStroke>,
        bounds: RectF,
        preContext: String,
    ): String

    companion object {
        /** Returned when recognition fails or the model is unavailable. */
        const val FALLBACK_TEXT = "unrecognized"
    }
}

File: apps/notesprout_flutter/lib/recognition/handwriting_recognizer.dart
import 'package:flutter/foundation.dart';

import '../domain/stroke.dart';
import 'mlkit_recognizer.dart';

/// Returned when recognition fails or the model is unavailable (mirrors native FALLBACK_TEXT).
const kFallbackText = 'unrecognized';

/// General-purpose handwriting recognition — operates on raw stroke data with no knowledge of the
/// caller (heading vs text). Port of the native `recognition/HandwritingRecognizer.kt`.
///
/// Implementations: [MlKitRecognizer] (Android/BOOX + iPad) and [UnsupportedRecognizer] (desktop —
/// ML Kit has no desktop build, so the feature is simply absent there).
abstract class HandwritingRecognizer {
  /// Whether recognition exists on this platform at all (false on desktop → hide the UI).
  bool get supported;

  /// Whether the model is downloaded and the recognizer is ready to run now.
  bool get isReady;

  /// Download/init the model if needed (idempotent). Returns true when ready.
  Future<bool> ensureReady();

  /// Recognize the handwriting in [strokes], returning the top candidate or [kFallbackText].
  /// [preContext] chains the previously recognized text for accuracy (page pipeline, Stage 2).
  Future<String> recognize(List<StrokeData> strokes, {String preContext = ''});

  void dispose();
}

/// No-op recognizer for platforms without ML Kit (desktop/web).
class UnsupportedRecognizer implements HandwritingRecognizer {
  @override
  bool get supported => false;
  @override
  bool get isReady => false;
  @override
  Future<bool> ensureReady() async => false;
  @override
  Future<String> recognize(List<StrokeData> strokes, {String preContext = ''}) async =>
      kFallbackText;
  @override
  void dispose() {}
}

HandwritingRecognizer _create() {
  switch (defaultTargetPlatform) {
    case TargetPlatform.android:
    case TargetPlatform.iOS:
      return MlKitRecognizer();
    default:
      return UnsupportedRecognizer();
  }
}

/// App-level singleton holding the active recognizer (mirrors native HandwritingRecognizerProvider).
class Recognition {
  Recognition._();
  static HandwritingRecognizer? _instance;
  static HandwritingRecognizer get instance => _instance ??= _create();
}
